### PR TITLE
media-gfx/solvespace: new package

### PR DESCRIPTION
**Pull request by ktrace**
Parametric 3D/2D CAD software with DXF import/export
 and STL export for 3D printing.

Signed-off-by: Victor Kustov <ktrace@yandex.ru>
Package-Manager: Portage-2.3.49, Repoman-2.3.10

### CODE REVIEW

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+DXFRW_PV="0.6.3"
+DXFRW_PN="libdxfrw"
+DXFRW_P="${DXFRW_PN}-${DXFRW_PV}-${DXFRW_COMMIT}"
+
+inherit cmake-utils
+
+DESCRIPTION="Parametric 2d/3d CAD"
+HOMEPAGE="http://solvespace.com"
+SRC_URI="https://github.com/solvespace/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz
+https://github.com/solvespace/${DXFRW_PN}/archive/${DXFRW_COMMIT}.zip -> ${DXFRW_P}.zip"
```
Comment (hydrapolic) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: A tab here would make it a bit more readable.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+LICENSE="GPL-3"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE=""
```
Comment (hydrapolic) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: You can drop it if it's empty.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+RDEPEND="dev-cpp/gtkmm:2.4
```
Comment (hydrapolic) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Adding slot operators for rebuild? As with json-c.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	media-libs/libpng:0=
+	virtual/opengl"
+DEPEND="${RDEPEND}
+	virtual/pkgconfig"
+
+GIT_COMMIT_HASH="4d1e1341d926ac356b295d5cc3d61c7a6cd7d07d"
+
+src_prepare() {
+	default
+	rm -rf "${S}/extlib/${DXFRW_PN}"
```
Comment (hydrapolic) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Missing || die

Comment (hydrapolic) [reply media-gfx/solvespace/solvespace-2.3.ebuild]: Do you really need -f? Do you need ${S}?

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# Copyright 1999-2018 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
+DXFRW_COMMIT="87ff1082f49b1286a033ac8f38fe8dde0975bb8b"
+DXFRW_PV="0.6.3"
+DXFRW_PN="libdxfrw"
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Please kill this one and just inline the name; putting it into variable has very little benefit and causes unnecessary indirection.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+DXFRW_P="${DXFRW_PN}-${DXFRW_PV}-${DXFRW_COMMIT}"
+
+inherit cmake-utils
+
+DESCRIPTION="Parametric 2d/3d CAD"
+HOMEPAGE="http://solvespace.com"
+SRC_URI="https://github.com/solvespace/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz
+	https://github.com/solvespace/${DXFRW_PN}/archive/${DXFRW_COMMIT}.zip -> ${DXFRW_P}.zip"
+
+LICENSE="GPL-3"
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Could you verify whether copyright notices on top of source files don't specify '[...] Public License version 3 or (at your option) newer' or something similar? If so, please make it `GPL-3+`.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+DXFRW_COMMIT="87ff1082f49b1286a033ac8f38fe8dde0975bb8b"
+DXFRW_PV="0.6.3"
+DXFRW_PN="libdxfrw"
+DXFRW_P="${DXFRW_PN}-${DXFRW_PV}-${DXFRW_COMMIT}"
+
+inherit cmake-utils
+
+DESCRIPTION="Parametric 2d/3d CAD"
+HOMEPAGE="http://solvespace.com"
+SRC_URI="https://github.com/solvespace/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Also please inline `${PN}` here, to make it easier to copy-paste the address to look at GitHub site.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	dev-libs/json-c:=
+	media-libs/fontconfig
+	media-libs/freetype:2
+	media-libs/glew:0=
+	media-libs/libpng:0=
+	virtual/opengl"
+DEPEND="${RDEPEND}
+	virtual/pkgconfig"
+
+GIT_COMMIT_HASH="4d1e1341d926ac356b295d5cc3d61c7a6cd7d07d"
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Is this really useful? I'd say it'll be really easy to miss updating it.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+DXFRW_PV="0.6.3"
+DXFRW_PN="libdxfrw"
+DXFRW_P="${DXFRW_PN}-${DXFRW_PV}-${DXFRW_COMMIT}"
+
+inherit cmake-utils
+
+DESCRIPTION="Parametric 2d/3d CAD"
+HOMEPAGE="http://solvespace.com"
+SRC_URI="https://github.com/solvespace/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz
+	https://github.com/solvespace/${DXFRW_PN}/archive/${DXFRW_COMMIT}.zip -> ${DXFRW_P}.zip"
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Why .zip and not .tar.gz?

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+GIT_COMMIT_HASH="4d1e1341d926ac356b295d5cc3d61c7a6cd7d07d"
+
+src_prepare() {
+	default
+	rm -r "extlib/${DXFRW_PN}" || die  "rm extlib/${DXFRW_PN} failed"
+	mv -T "${WORKDIR}/${DXFRW_PN}-${DXFRW_COMMIT}" "extlib/${DXFRW_PN}" || die "move ${DXFRW_PN}-${DXFRW_COMMIT} failed"
+	# fixed in master already
+	sed -i '/include(GetGitCommitHash)/d' CMakeLists.txt || die 'remove GetGitCommitHash by sed failed'
+	sed -i 's/CHAR_WIDTH/CHAR_WIDTH_/' src/textwin.cpp || die 'sed failed'
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: You can pass multiple files to `sed`. Also use `sed -i -e '...' files` for consistency.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+	virtual/opengl"
+DEPEND="${RDEPEND}
+	virtual/pkgconfig"
+
+GIT_COMMIT_HASH="4d1e1341d926ac356b295d5cc3d61c7a6cd7d07d"
+
+src_prepare() {
+	default
+	rm -r "extlib/${DXFRW_PN}" || die  "rm extlib/${DXFRW_PN} failed"
+	mv -T "${WORKDIR}/${DXFRW_PN}-${DXFRW_COMMIT}" "extlib/${DXFRW_PN}" || die "move ${DXFRW_PN}-${DXFRW_COMMIT} failed"
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: I suppose there's no chance to split that into separate package?

File: media-gfx/solvespace/solvespace-2.3.ebuild (hunk 6 above)
Comment (mgorny) [reply media-gfx/solvespace/solvespace-2.3.ebuild]: Nah. Unless you can find somewhere where program authors explicitly say 'or newer version', let's leave it as GPL-3.

File: media-gfx/solvespace/solvespace-2.3.ebuild (hunk 11 above)
Comment (ktrace) [reply media-gfx/solvespace/solvespace-2.3.ebuild]: Fork of dead lib, specific patches, no tags or releases... Forever 9999... I think no, till upstream release or tag it.

File: media-gfx/solvespace/solvespace-2.3.ebuild (hunk 8 above)
Comment (ktrace) [reply media-gfx/solvespace/solvespace-2.3.ebuild]: I think it's indifferent in this case. 

> src_configure() {
> ...
>         -DGIT_COMMIT_HASH=${GIT_COMMIT_HASH}

It's same challenge to attantion like  

> src_configure() {
> ...
>         -DGIT_COMMIT_HASH="4d1e1341d926ac356b295d5cc3d61c7a6cd7d07d"

Comment (mgorny) [reply media-gfx/solvespace/solvespace-2.3.ebuild]: Let me rephrase. Is there any value in passing `-DGIT_COMMIT_HASH` at all? Can you skip it, leave it empty or something like that?

Comment (ktrace) [reply media-gfx/solvespace/solvespace-2.3.ebuild]: Or value pass via -D or patch CMakeLists.txt with set this variable. It cannot be empty or skipped at all.

Comment (mgorny) [reply media-gfx/solvespace/solvespace-2.3.ebuild]: Can you put `${PV}` there instead? That will at least update itself automatically.

Comment (mgorny) [reply media-gfx/solvespace/solvespace-2.3.ebuild]: Ok, so maybe move it to the top and add comment saying you need to update it on bump.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	virtual/opengl"
+DEPEND="${RDEPEND}
+	virtual/pkgconfig"
+
+GIT_COMMIT_HASH="4d1e1341d926ac356b295d5cc3d61c7a6cd7d07d"
+
+src_prepare() {
+	default
+	rm -r "extlib/libdxfrw" || die  "rm extlib/libdxfrw failed"
+	mv -T "${WORKDIR}/libdxfrw-${DXFRW_COMMIT}" "extlib/libdxfrw" || die "move libdxfrw-${DXFRW_COMMIT} failed"
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: `mv -T` is not portable.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+	sed -i '/include(GetGitCommitHash)/d' CMakeLists.txt || die 'remove GetGitCommitHash by sed failed'
+	# fixed in master already
+	sed -i -e 's/CHAR_WIDTH/CHAR_WIDTH_/' src/textwin.cpp src/toolbar.cpp src/win32/w32main.cpp src/ui.h || die 'sed failed'
+}
+
+src_configure() {
+	local mycmakeargs=(
+		-DGIT_COMMIT_HASH=${GIT_COMMIT_HASH}
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Quote it, just in case.

File: media-gfx/solvespace/solvespace-2.3.ebuild [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+src_configure() {
+	local mycmakeargs=(
+		-DGIT_COMMIT_HASH="${GIT_COMMIT_HASH}"
+	)
+	cmake-utils_src_configure
+}
+
+pkg_postinst() {
+	gnome2_icon_cache_update
```
Comment (mgorny) [inline media-gfx/solvespace/solvespace-2.3.ebuild]: Works without inheriting? ;-P
